recipes/full_finetune_distributed.py

```diff
@@ -259,6 +267,10 @@ def _setup_model(
             ),
         )
 
+        if model_tie:  # Tie the weights of the model if required
+            model.output.weight = model.tok_embeddings.weight
```

Comment:
It was pointed out by @rohan-varma that this may not actually do what we expect because FSDP has already sharded the params, so let's double-confirm via testing that the weights are tied correctly here.

Reply:
Thank you for pointing out it, I check the model weight after training, they are not same. Is there any solution? I'm not familiar with that. This can cause some problems if the weights are tied before FSDP. [issue](https://github.com/pytorch/torchtune/issues/616#issuecomment-2028619062)

Reply:
OK sorry for the back and forth on this. Confirmed with @rohan-varma that we should not tie weights after FSDP wrapping after all. The main issue was not FSDP but the initialization on meta device. Unfortunately, weight tying + meta device is tricky because the usage of `to_empty` breaks existing references. 

Instead, for Gemma we can do everything on CPU without using meta device at all, basically initializing the model on CPU for every rank and then defining a more vanilla FSDP without the `param_init_fn` we currently have. This should work fine for smaller models (at least up to 7B). @kartikayk put together a snippet on what this can look like, you can find it [here](https://gist.github.com/ebsmothers/332b8432c19c0717778e553b133face3).

We need to decide what the best way to expose this is, but for now feel free to create a separate recipe for Gemma. It should look pretty much the same as the existing `full_finetune_distributed.py`, but with the changes needed to initialize everything on CPU and perform weight tying there before wrapping with FSDP.

Thanks also to @awgu for helping debug this.

Reply:
Thank you for your suggestion, it works well.